Sources/Support/StoreMessagesHelper.swift

```diff
@@ -0,0 +1,195 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  StoreMessagesHelper.swift
+//
+//  Created by Antonio Rico Diez on 27/9/23.
+
+import StoreKit
+
+class StoreMessagesHelper {
+
+    private let systemInfo: SystemInfo
+    private let showStoreMessagesAutomatically: Bool
+    private let storeMessagesProvider: StoreMessagesProvider
+
+    private var deferredMessages: [StoreMessage] = []
+
+    init(systemInfo: SystemInfo,
+         showStoreMessagesAutomatically: Bool) {
+        self.systemInfo = systemInfo
+        self.showStoreMessagesAutomatically = showStoreMessagesAutomatically
+        self.storeMessagesProvider = StoreMessagesProviderWrapper()
+    }
+
+    @available(iOS 16.0, *)
+    init(systemInfo: SystemInfo,
+         showStoreMessagesAutomatically: Bool,
+         storeMessagesProvider: StoreMessagesProvider = StoreMessagesProviderWrapper()) {
+        self.systemInfo = systemInfo
+        self.showStoreMessagesAutomatically = showStoreMessagesAutomatically
+        self.storeMessagesProvider = storeMessagesProvider
+    }
+
+    #if os(iOS) || targetEnvironment(macCatalyst) || VISION_OS
+
+    @available(iOS 16.4, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    func deferMessagesIfNeeded() async throws {
+        guard !self.showStoreMessagesAutomatically else {
+            return
+        }
+        // Compilation fails in older xcode versions with a similar issue to https://github.com/apple/swift/issues/57379
+        #if swift(>=5.8)
+        Task(priority: .medium) { [weak self] in
+            guard let storeMessagesProvider = self?.storeMessagesProvider else {
+                return
+            }
+            for try await message in storeMessagesProvider.messages {
+                await MainActor.run { [weak self] in
+                    self?.deferredMessages.append(message)
+                }
+            }
+        }
+        #else
+        Logger.error("Defer messages are not enabled when compiling using Xcode <14.2")
+        #endif
+    }
+
+    @available(iOS 16.4, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    @MainActor
+    func showStoreMessages(types: Set<StoreMessageType>) {
+        for message in self.deferredMessages {
+            if let messageType = message.reason.messageType,
+               types.contains(messageType) {
+                    do {
+                        try message.display(in: self.systemInfo.currentWindowScene)
```

Comment:
Have you tried displaying multiple messages? My bet is that the second one would fail. I wonder if we should just display the first one only.

Reply:
Made this change here: https://github.com/RevenueCat/purchases-ios/pull/3252/commits/7a8586f8d478fc85411f247b4f3d7fea1708613b

Reply:
@NachoSoto could you elaborate? Why would you expect the second one to fail? 

Apple's docs suggest this as the way to do it https://developer.apple.com/documentation/storekit/displaymessageaction

And it seems like others are already doing it like that
https://github.com/brave/brave-ios/blob/509191082e749fd6e96b87e3172b4528e4eabc98/Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController%2BCallout.swift#L273

Reply:
Reverted the changes so we display all messages again: https://github.com/RevenueCat/purchases-ios/pull/3252/commits/b65d22d5c35ee3732db27a5b8c580526e7983a5c

Reply:
I think that's the safest unless we have a clear reason why we know it would fail. This is pretty hard to test in practice, so following Apple's example as close as we can seems like a good bet